horovod/torch/__init__.py

```diff
@@ -129,15 +131,43 @@ def broadcast_parameters(params, root_rank):
     `model.named_parameters()`, or `model.parameters()`.
 
     Arguments:
-        params: The list of parameters to broadcast.
+        params: One of the following:
+            - list of parameters to broadcast
+            - dict of parameters to broadcast
+            - torch.optim.Optimizer whose state will be broadcast
         root_rank: The rank of the process from which parameters will be
                    broadcasted to all other processes.
     """
     if isinstance(params, dict):
         params = sorted(params.items())
-    else:
+    elif isinstance(params, list):
         # support both named_parameters() and regular parameters()
         params = [p if isinstance(p, tuple) else (None, p) for p in params]
+    elif isinstance(params, torch.optim.Optimizer):
+        # Newly created optimizers will not have their state initialized, so
+        # do that initialization here
+        if len(params.state_dict()['state']) == 0:
+            for group in params.param_groups:
+                for p in group['params']:
+                    p.grad = torch.autograd.Variable(
+                        p.data.new(p.size()).zero_())
+            params.step()
+
+        new_params = []
+        occurrences = collections.defaultdict(int)
+        for group in params.state_dict()['param_groups']:
```

Comment:
Is it possible that these groups will be ordered differently on different nodes?

Reply:
No, I'll add a comment on that.  The `params` list is ordered by the order of the layers in the model.